addons/wallet/models/wallet.py

```diff
@@ -0,0 +1,105 @@
+    _name = "wallet"
+    _description = "Partner Wallet"
+
+    currency_id = fields.Many2one('res.currency',
+                                  string='Currency',
+                                  readonly=True,
+                                  default=lambda self: self.env.company.currency_id.id,
+                                  required=True)
+
+    partner_id = fields.Many2one("res.partner", required=True)
+
+    wallet_transaction_ids = fields.One2many("wallet.transaction", "wallet_id")
+    wallet_transactions_count = fields.Integer(compute="_compute_count_wallet_transaction")
+    balance = fields.Monetary(currency_field='currency_id', compute="_compute_balance")
+    actual_balance = fields.Monetary(currency_field='currency_id', compute="_compute_actual_balance")
+
+    @api.depends("wallet_transaction_ids")
+    def _compute_count_wallet_transaction(self):
+        for record in self:
+            record.wallet_transactions_count = len(record.wallet_transaction_ids)
+
+    @api.depends("wallet_transaction_ids")
+    def _compute_balance(self):
+        for record in self:
+            record.balance = reduce(
+                lambda o1, o2: (o1 + o2.get(currency_id=record.currency_id)),
+                record.wallet_transaction_ids.filtered(lambda w_t: not w_t.is_debit or w_t.is_confirmed), 0
+            )
+
+    @api.depends("wallet_transaction_ids")
+    def _compute_actual_balance(self):
+        for record in self:
+            record.actual_balance = reduce(
+                lambda o1, o2: (o1 + o2.get(currency_id=record.currency_id)),
+                record.wallet_transaction_ids, 0
+            )
+
+    def add_debit(self, order_id):
+        can_be_retained = self.how_much_can_be_retained(order_id.get_total_amount(), order_id.currency_id)
+        if order_id.add_update_wallet_transaction(can_be_retained):
+            self.release_no_confirmed_order()
+            self.env["wallet.transaction"].create({
+                'is_debit': True,
+                'wallet_id': self.id,
+                'sale_order_line_id': order_id.order_line_wallet_id.id
+            })
+
+    def add_credit(self, gift_card_id):
+        self.env["wallet.transaction"].create({
+            'is_debit': False,
+            'wallet_id': self.id,
+            'gift_card_id': gift_card_id.id
+        })
+
+    def how_much_can_be_retained(self, amount, currency_id):
+        return min(amount, self.currency_id._convert(self.balance, currency_id, self.env.company, fields.Date.today()))
+
+    def release_no_confirmed_order(self):
+        self.wallet_transaction_ids.filtered(lambda w_t: w_t.is_debit and not w_t.is_confirmed).unlink()
+
+    def release_amount(self, wallet_transaction_id):
+        self.wallet_transaction_ids -= wallet_transaction_id
+
+
+class WalletTransaction(models.Model):
+    _name = "wallet.transaction"
+    _description = "Partner Wallet Transaction"
+
+    is_debit = fields.Boolean(help="Is a debit transaction?.")
```

Comment:
It seems duplicated stored field...

is_debit = bool(self.sale_order_line_id)
is_debit = not bool(self.gift_card_id)

Reply:
Weak coupling.
it is the Boolean that decide whether it is a debit or a credit.

Reply:
weak ? but useless imo

It will just bring a way to have ambiguous data.

Use at least a compute method if really needed